File: plugins/com.googlecode.contraildb.core/src/com/googlecode/contraildb/core/IPreparedQuery.java
package com.googlecode.contraildb.core;

import java.io.IOException;
import java.util.List;

public interface IPreparedQuery<T extends Item>  {
	
	/**
	 * @return The session that created this query 
	 */
	public IContrailSession getSession();

	public List<T> list(FetchOptions fetchOptions) throws IOException;

	public List<T> list() throws IOException;

	public Iterable<T> iterate(FetchOptions fetchOptions) throws IOException;

	public Iterable<T> iterate() throws IOException;

	public T item() throws IOException;

	public int count() throws IOException;

	public Iterable<Identifier> identifiers() throws IOException;
}

File: ContrailNet/ContrailNet/Contrail/IPreparedQuery.cs
using System;
using System.Collections.Generic;

namespace Contrail {

public interface IPreparedQuery<T>  where T : Item
{
    // The session that created this query 
    string Session { get; set; }

	// Enumerates the identifiers of all matching items.
	// This method id the most basic way to get results and is the fastest.
	IEnumerator<Identifier> identifiers();

	// A convenience method that returns items instead of identifiers.
	IEnumerator<T> iterate(FetchOptions fetchOptions);

	// A convenience method that returns items instead of identifiers.
	IEnumerator<T> iterate();

	// A convenience method that returns a list of all matching items
	List<T> List(FetchOptions fetchOptions);

	// A convenience method that returns a list of all matching items
	List<T> List();

	// A convenience method that returns the first matching item
	T Item();

	// A convenience method that returns just the number of matching items
	int count();
}

}